myia/anf_ir.py

```diff
@@ -0,0 +1,317 @@
+    can be stored in this class.
+
+    Attributes:
+        ast: The AST node that created this graph.
+        name: The function name.
+
+    """
+
+    ast: AST
+    name: str
+
+
+class ANFNode(Node):
+    """A node in the graph-based ANF IR.
+
+    There are four types of nodes: Function applications; parameters; return
+    values; and constants such as numbers and functions.
+
+    Attributes:
+        inputs: If the node is a function application, the first node input is
+            the function to apply, followed by the arguments. These are use-def
+            edges.
+        value: The value of this node, if it is a constant. Parameters have the
+            special value `PARAMETER`.
+        graph: The function definition graph that this node belongs to for
+            values and parameters.
+        uses: A set of tuples with the nodes that use this node alongside with
+            the index. These def-use edges are the reverse of the `inputs`
+            attribute, creating a doubly linked graph structure.
+        debug: A dictionary with debug information about this node e.g. a
+            human-readable name and the Python source code.
+
+    """
+
+    def __init__(self, inputs: Iterable['ANFNode'], value: Any,
+                 graph: Graph) -> None:
+        """Construct a node."""
+        self._inputs = Inputs(self, inputs)
+        self.value = value
+        self.graph = graph
+        self.uses: Set[Tuple[ANFNode, int]] = set()
+        self.debug = NodeDebug()
+
+    @property
+    def inputs(self) -> 'Inputs':
+        """Return the list of inputs."""
+        return self._inputs
+
+    @inputs.setter
+    def inputs(self, value: Iterable['ANFNode']) -> None:
+        """Set the list of inputs."""
+        self._inputs.clear()  # type: ignore
+        self._inputs = Inputs(self, value)
+
+    @property
+    def incoming(self) -> Iterable['ANFNode']:
+        """Return incoming nodes in order."""
+        return iter(self.inputs)
+
+    @property
+    def outgoing(self) -> Iterable['ANFNode']:
+        """Return uses of this node in random order."""
+        return (node for node, index in self.uses)
+
+    def __copy__(self) -> 'ANFNode':
+        """Copy this node.
+
+        This method is used by the `copy` module. It ensures that copied nodes
+        will have correct `uses` information. Debug information is not copied.
+
+        """
+        cls = self.__class__
+        obj = cls.__new__(cls)  # type: ignore
+        ANFNode.__init__(obj, self.inputs, self.value, self.graph)
+        return obj
+
+    def replace(self, other: 'ANFNode') -> None:
+        """Replace one node in the graph with another.
+
+        Args:
+            other: The node to replace this one with. All incoming and outgoing
+                edges will be replace.
+
+        """
+        other.inputs = self.inputs
+        self.inputs.clear()  # type: ignore
```

Comment:
I'm not sure this is worth doing. After being replaced, this node is dead, so you want to remove uses of it, but in doing so you may kill other nodes, and you might need to walk through the graph recursively to prune the whole dead subtree. Either we do exactly that, or we can leave it to another pass.

Reply:
Doing a full graph walk at each step might be less efficient than removing a whole set of nodes one by one, and then doing one pruning pass in the end, so I don't want to enforce it here. I can just get rid of this method and we can introduce some helper functions in the optimization framework later on.